### PR TITLE
Fix OSX x86 warnings.

### PR DESCRIPTION
**Pull request by jaykrell**
In particular:

 1) Be sure to pass a function all of its parameters, no more, no less.
 You can't have a stack imbalance because parameters usually fit in registers,
 and because calling conventions except for __stdcall are explicitly designed
 to accomodate common programming mistakes, e.g. printf ("%d") and printf ("%d", 1, 2),
 are well defined on all modern ABIs to never crash (except for the extreme
 case of printf ("%d") where stack is near the top and it accesses invalid next page,
 most unlikely).

 2) size_t and gsize are not always the same type.
 For example, one can be unsigned int and the other unsigned long (32bit).
 Or unsigned long and unsigned long long (64bit).

 3) gsize is not necessarily unsigned long, nor even the same size as it.
 It could be any of:
  unsigned int -- viable on all 32bit (Win32 and Unix32)
  unsigned long -- viable on Win32 and all Unix (Unix32 and Unix64)
  unsigned long long -- viable on Win64 and Unix64
  unsigned __int64 -- viable on Win64, possibly same as unsigned long long

 To printf a gsize in the "u" style (not the "x" nor "d" styled), G_GSIZE_FORMAT is provided.
  (Fix for Win32 and Win64 forthcoming.)

 We should strongly consider typedef size_t gsize but this a larger change.

 4) Pop without push.

```
mono-publib.c:27:30: warning: incompatible function pointer types initializing 'gpointer (*)(gsize)' (aka 'void *(*)(unsigned int)') with an expression of type
      'void *(*)(size_t)' (aka 'void *(*)(unsigned long)') [-Wincompatible-function-pointer-types]
        GMemVTable g_mem_vtable = { vtable->malloc, vtable->realloc, vtable->free, vtable->calloc};
                                    ^~~~~~~~~~~~~~
mono-publib.c:27:46: warning: incompatible function pointer types initializing 'gpointer (*)(gpointer, gsize)' (aka 'void *(*)(void *, unsigned int)') with an
      expression of type 'void *(*)(void *, size_t)' (aka 'void *(*)(void *, unsigned long)') [-Wincompatible-function-pointer-types]
        GMemVTable g_mem_vtable = { vtable->malloc, vtable->realloc, vtable->free, vtable->calloc};
                                                    ^~~~~~~~~~~~~~~
mono-publib.c:27:77: warning: incompsatible function pointer types initializing 'gpointer (*)(gsize, gsize)' (aka 'void *(*)(unsigned int, unsigned int)') with an
      expression of type 'void *(*)(size_t, size_t)' (aka 'void *(*)(unsigned long, unsigned long)') [-Wincompatible-function-pointer-types]
        GMemVTable g_mem_vtable = { vtable->malloc, vtable->realloc, vtable->free, vtable->calloc};
                                                                                   ^~~~~~~~~~~~~~
mono-time.c:204:1: warning: unused label 'exit' [-Wunused-label]
exit:
^~~~~

mono-threads-posix.c:92:43: warning: incompatible pointer types passing 'gsize *const' (aka 'unsigned int *const') to parameter of type 'size_t *'
      (aka 'unsigned long *') [-Wincompatible-pointer-types]
                res = pthread_attr_getstacksize (&attr, stack_size);
                                                        ^~~~~~~~~~
/usr/include/pthread.h:250:85: note: passing argument to parameter here
int pthread_attr_getstacksize(const pthread_attr_t * __restrict, size_t * __restrict);
                                                                                    ^

profiler.c:1062:57: warning: incompatible pointer types passing 'void (MonoProfiler *, MonoProfilerGCEvent, uint32_t)' (aka 'void (struct _MonoProfiler *,
      MonoProfilerGCEvent, unsigned int)') to parameter of type 'MonoProfilerGCEvent2Callback' (aka 'void (*)(struct _MonoProfiler *, MonoProfilerGCEvent,
      unsigned int, int)') [-Wincompatible-pointer-types]
                mono_profiler_set_gc_event_callback (current->handle, gc_event_cb);
                                                                      ^~~~~~~~~~~
../../mono/metadata/profiler-events.h:85:1: note: passing argument to parameter 'cb' here
MONO_PROFILER_EVENT_3(gc_event, GCEvent2, MonoProfilerGCEvent, event, uint32_t, generation, mono_bool, is_serial)
^
profiler.c:908:2: note: expanded from macro 'MONO_PROFILER_EVENT_3'
        _MONO_PROFILER_EVENT(name, type)
        ^
profiler.c:897:103: note: expanded from macro '_MONO_PROFILER_EVENT'
        mono_profiler_set_ ## name ## _callback (MonoProfilerHandle handle, MonoProfiler ## type ## Callback cb) \
                                                                                                             ^
marshal.c:4770:72: warning: format specifies type 'unsigned long' but the argument has type 'gsize' (aka 'unsigned int') [-Wformat]
                mono_error_set_out_of_memory (error, "Could not allocate %lu bytes", size);
                                                                         ~~~         ^~~~
                                                                         %u

trace.c:146:32: warning: unknown warning group '-Wframe-address', ignored [-Wunknown-warning-option]
                               ^
trace.c:148:24: warning: pragma diagnostic pop could not pop, no matching push [-Wunknown-pragmas]
                       ^

```

### CODE REVIEW

**Comment by vargaz** [comment]
Afaik gsize is supposed to be the same as size_t, from the glib docs:

An unsigned integer type of the result of the sizeof operator, corresponding to the size_t type defined in C99.

**Comment by jaykrell** [comment]
It is the same size but not the same type. size and type are not the same.

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by jaykrell** [comment]
Let’s fix gsize then.
https://github.com/mono/mono/pull/8570

File: mono/utils/mono-time.c [hunk 1]
```diff
@@ -201,6 +201,7 @@ mono_msec_boottime (void)
     retval = (tv.tv_sec * tccSecondsToMillieSeconds) + (tv.tv_usec / tccMillieSecondsToMicroSeconds);
 
 #endif /* HAVE_CLOCK_MONOTONIC */
+	goto exit; // fix ifdef-dependent warning about unused label
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/utils/mono-time.c]: Sure.

**Comment by lewurm** [comment]
@monojenkins squash

**Comment by lewurm** [comment]
@monojenkins build Linux AArch64

**Comment by lewurm** [comment]
@monojenkins build Linux x64

**Comment by lewurm** [comment]
@monojenkins build Linux ARMv5

**Pull request merged**